fvm/transactionInvoker.go

```diff
@@ -155,6 +151,72 @@ func (executor *transactionExecutor) Execute() error {
 	return nil
 }
 
+func (executor *transactionExecutor) Preprocess() error {
+	if !executor.TransactionBodyExecutionEnabled {
+		return nil
+	}
+
+	return executor.handleError(
+		executor.PreprocessTransactionBody(),
+		"preprocessing")
+}
+
+func (executor *transactionExecutor) Execute() error {
+	return executor.handleError(executor.execute(), "executing")
+}
+
+func (executor *transactionExecutor) PreprocessTransactionBody() error {
+	meterParams, err := getBodyMeterParameters(
+		executor.ctx,
+		executor.proc,
+		executor.txnState,
+		executor.derivedTxnData)
+	if err != nil {
+		return fmt.Errorf("error gettng meter parameters: %w", err)
+	}
+
+	txnId, err := executor.txnState.BeginNestedTransactionWithMeterParams(
```

Comment:
What is "nested" transaction?

If block 5 has 2 collections, collectionA and collection B. CollectionA has 3 transactions: tx1, tx2, tx3. And CollectionB has 2 transactions: tx4, and tx5.

Does it mean, tx1, tx2, tx3, tx4, tx5 are "nested"?

Reply:
https://en.wikipedia.org/wiki/Nested_transaction

Reply:
Thanks for the reference.

But it's not clear to me, why it's called "nested" transaction.

In my above example, where the block has 2 collections and 5 transactions in total. A nested transaction is nested if there is a host transaction. What would be the host transaction? 

If there is no host transaction, why do we call these transactions "nested"?

Reply:
A nested transaction has nothing to do with block/collection.  A nested transaction is literally a transaction nested/scoped inside another/parent transaction.  it allows you to partially rollback/undo changes the parent transaction without aborting it.

In the fvm case, the "host/main" transaction is everything executed by TransactionProcedure.  Under normal condition, we want to commit everything.  But when a user error occurs during transaction body execution, we want to rollback all the changes make by the user, but continue to deduct fees.  To achieve this behavior, a nested transaction is created for the "transaction body execution" part of the transaction, and this nested transaction is only commit when there no error.